File: Python/add_strings.py
# https://leetcode.com/problems/add-strings/description/
from utils.examination import Examiner, Testcase, Logger


class Solution:
    def addStrings(self, num1: str, num2: str) -> str:
        num1_i: int = len(num1) - 1
        num2_i: int = len(num2) - 1
        p: int = 0
        sum_: str = ''
        while num1_i >= 0 or num2_i >= 0:
            n1: int = 0
            if 0 <= num1_i:
                n1 = ord(num1[num1_i]) - ord('0')
            n2: int = 0
            if 0 <= num2_i:
                n2 = ord(num2[num2_i]) - ord('0')
            s: int = n1 + n2 + p
            if 10 <= s:
                p = 1
                s = s - 10
            else:
                p = 0
            sum_ = chr(s + ord('0')) + sum_
            num1_i -= 1
            num2_i -= 1
        if 0 < p:
            sum_ = chr(p + ord('0')) + sum_
        return sum_

def examine()-> None:
    Examiner(Solution(), 'addStrings').run((
        Testcase(('11', '123'), '134'),
        Testcase(('456', '77'), '533'),
        Testcase(('0', '0'), '0'),
        Testcase(('32', '0'), '32'),
        Testcase(('999', '1'), '1000'),
    ))

File: TypeScript/solutions/add-strings.ts
// https://leetcode.com/problems/add-strings/description/
import { Examiner, Testcase } from '../utils';


function addStrings(num1: string, num2: string): string {
    let num1I = num1.length - 1;
    let num2I = num2.length - 1;
    let p = 0;
    let sum = '';
    while (num1I >= 0 || num2I >= 0) {
        const n1 = num1I >= 0 ? num1.charCodeAt(num1I) - '0'.charCodeAt(0) : 0;
        const n2 = num2I >= 0 ? num2.charCodeAt(num2I) - '0'.charCodeAt(0) : 0;
        let s = n1 + n2 + p;
        if (s >= 10) {
            p = 1;
            s -= 10;
        } else {
            p = 0;
        }
        sum = String.fromCharCode(s + '0'.charCodeAt(0)) + sum;
        num1I--;
        num2I--;
    }
    if (p > 0) sum = String.fromCharCode(p + '0'.charCodeAt(0)) + sum;
    return sum;
}

export function examine(): void {
    new Examiner(addStrings).run([
        new Testcase(['11', '123'], '134'),
        new Testcase(['456', '77'], '533'),
        new Testcase(['0', '0'], '0'),
        new Testcase(['32', '0'], '32'),
        new Testcase(['999', '1'], '1000'),
    ]);
}